### ISSUE TITLE
TypeError: Cannot assign to read only property 'componentDidUpdate' of object

### ISSUE THREAD

**Issue by bjrnt** [mobxjs/mobx-react#231]
It took me ages to figure out what the cause was, so I'm just going to upload this issue here in case other people are googling for the same thing.

This is caused by `mobx-react@4.1.4`, and you can find the details [here](https://github.com/mobxjs/mobx-react/commit/43d4d71e00f0a5f7c44f097dd2c99ae9e4ea7c7b). When we updated to this version a bunch of our components that were not using these best practices suddenly exploded and caused a massive headache. :(

**Comment by mweststrate** [comment]
Note it is also mentioned in the changelog: https://github.com/mobxjs/mobx-react/blob/master/CHANGELOG.md#414. Also note that without this fix, (so on older mobx-react versions), the components probably didn't become observers at all.

Nonetheless, sorry for the inconvenience it caused!

**Comment by wangyu0829** [comment]
I update to 4.1.5, there's a problem when i use react-hot-loader at the same time. 
**Uncaught TypeError: Cannot assign to read only property 'componentWillMount' of object**
how can I fix it，thx

**Comment by andykog** [comment]
@wangyu0829, I suspect, you are defining property like this:
```js
class extends React.Component {
  componentWillMount = () => { /**/ }
}
```
You should instead always define those methods on prototype:
```js
class extends React.Component {
  componentWillMount() { /**/ }
}
```

**Comment by wangyu0829** [comment]
@andykog  thx , but I have not define the componentWillMount, like this
```
@observer
export default class App extends Component {
    static propTypes = {
        appState: PropTypes.shape({})
    }

    static contextTypes = {
        intl: PropTypes.object
    }

    render() {
        return (
            <div className="page">
                <EntranceList />
            </div>
        )
    }
}
```

**Comment by drewpc** [comment]
I'm having the same issue.  I have not defined `componentWillMount` at all.  My class looks like this:

```
@observer
export default class MessageBlock extends React.Component {
    static propTypes = {
        blockObj: PropTypes.object.isRequired,
    };

    static defaultProps = {
        sizeX: 1,
        sizeY: 2,
    };

    static defaultClassNames = [
        'bs-block-has-padding',
    ];

    render() {
        const { blockObj} = this.props;

        let item = {};

        if (blockObj.items.length > 0) {
            item = blockObj.items[0];
        }

        return (
            <div>
                <MessageBlockItem itemObj={item} />
            </div>
        );
    }
};
```
And I get the following error:
```
common.bundle.js:formatted:9693 Uncaught TypeError: Cannot assign to read only property 'componentWillMount' of object '#<MessageBlock>'
    at patchProperty (http://0.0.0.0:3001/common.bundle.js:8681:17)
    at http://0.0.0.0:3001/common.bundle.js:8712:10
    at Array.forEach (native)
    at http://0.0.0.0:3001/common.bundle.js:8711:52
    at Array.forEach (native)
    at reconcileWithStoredPrototypes (http://0.0.0.0:3001/common.bundle.js:8710:22)
    at assimilatePrototype (http://0.0.0.0:3001/common.bundle.js:8723:6)
    at patchReactClass (http://0.0.0.0:3001/common.bundle.js:8647:6)
    at Object.makeHot (http://0.0.0.0:3001/common.bundle.js:8600:13)
    at makeExportsHot (http://0.0.0.0:3001/common.bundle.js:57650:27)
```

**Comment by mweststrate** [comment]
Reopened. RHL breaks stuff like usual as it tries to modify the frozen prototype, but too much used to ignore :(

**Comment by mweststrate** [comment]
Published 4.1.8 where this check introduced in 4.1.4 is disabled again :-(

**Comment by drewpc** [comment]
thank you!
